src/transformers/models/vits/tokenization_vits.py

```diff
@@ -0,0 +1,137 @@
+# coding=utf-8
+# Copyright 2023 The MMS-TTS Authors and the HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tokenization class for VITS."""
+
+
+import json
+import os
+from typing import List, Optional, Tuple
+
+from ...tokenization_utils import PreTrainedTokenizer
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+VOCAB_FILES_NAMES = {"vocab_file": "vocab.json"}
+
+PRETRAINED_VOCAB_FILES_MAP = {
+    "vocab_file": {
+        "Matthijs/mms-tts-eng": "https://huggingface.co/Matthijs/mms-tts-eng/resolve/main/vocab.json",
+    }
+}
+
+PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES = {
+    # This model does not have a maximum input length.
+    "Matthijs/mms-tts-eng": 4096,
+}
+
+
+class VitsMmsTokenizer(PreTrainedTokenizer):
+    """
+    Construct a VITS tokenizer for MMS-TTS.
+
+    This tokenizer inherits from [`PreTrainedTokenizer`] which contains most of the main methods. Users should refer to
+    this superclass for more information regarding those methods.
+
+    Args:
+        vocab_file (`str`):
+            Path to the vocabulary file.
+        language (`str`, *optional*):
+            Language identifier.
+        add_blank (`bool`, *optional*, defaults to `True`):
+            Whether to insert token id 0 in between the other tokens.
+    """
+
+    vocab_files_names = VOCAB_FILES_NAMES
+    pretrained_vocab_files_map = PRETRAINED_VOCAB_FILES_MAP
+    max_model_input_sizes = PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES
+    model_input_names = ["input_ids", "attention_mask"]
+
+    def __init__(
+        self,
+        vocab_file,
+        pad_token="<pad>",
+        unk_token="<unk>",
+        language=None,
+        add_blank=True,
+        **kwargs,
+    ) -> None:
+        super().__init__(pad_token=pad_token, unk_token=unk_token, language=language, add_blank=add_blank, **kwargs)
+
+        with open(vocab_file, encoding="utf-8") as vocab_handle:
+            self.encoder = json.load(vocab_handle)
+
+        # The original model pads with token_id 0, but that doesn't work here
```

Comment:
Hey @ArthurZucker - wonder if you know a more elegant workaround for this tokenization issue?

Reply:
I think that the vocab should be updated in the `convert_original_checkpoint`, and just add the special tokens here. 
I am not sure I understand the problem, they pad with `0` but `0` maps to `"k"`. So they pad with `"k"`?

Reply:
Yes, they pad with token_id 0. The problem is that the tokenizer splits the input on the padding token and tokenizes these subsequences separately, so anywhere there is a `k` in the text it will disappear.

For example, `It stinks here` gets tokenized as `It stin` and `s here` because `k` is not treated as a regular token but as padding.

Reply:
I think @amyeroberts raised very valid questions. 
- if they pad with `k` then `k` is the padding token non? 
- if `<pad>` and `k` both point to `0`, we don't really have a way in `transformers` to make sure this can be done. But I don't see why it should be done since they pad with k! 
- if the problem is that our `tokenize` methods splits on the special tokens, including padding, but you don't want this to happen, then you can just overwrite `tokenize`. Given that this model does not seem to use `bos` nor `eos` you can probably ignore the lstrip and rstip logics, and just take what you have in `_tokenize`.

Reply:
@amyeroberts's explanation is correct: the model is trained with `0` as a padding token, and this also corresponds to the character `k`.

The text is first split at each character, and `<pad>` tokens inserted in the gaps. The chars are then tokenised to give the input ids. This is easiest illustrated with a few examples.

With no `k` input:
```
input: cat
split: <pad> c <pad> a <pad> t <pad>
tokens ids: [ 0, 12,  0, 26,  0, 33,  0]
```

With a `k` in the input:
```
input: kat
split: <pad> k <pad> a <pad> t <pad>
tokens ids: [ 0,  0,  0, 26,  0, 33,  0]
```

-> here we see how the text is first split, padding inserted, and then tokenised with the mapping `<pad> -> 0` and also `k -> 0`.

> We filter any token_ids corresponding to "<pad>" in our model?

This is **not** done - the attention layer in VITS uses relative positional representations (instead of absolute positional encodings). I presume that this layer learns to ignore any tokens on even indices that correspond to padding (0, 2, 4, ...), and only attends to ids on odd indices that correspond to the text input (1, 3, 5, ...). Note that the only way of verifying this would be to look at an attention plot or inspect the relative positional representation weights

Reply:
Now answering @ArthurZucker's points! 

> if they pad with `k` then `k` is the padding token non?

Explained above - they pad with the `0` token, which also corresponds to `k` in the vocab

> if `<pad>` and `k` both point to `0`, we don't really have a way in `transformers` to make sure this can be done

It is indeed the case that both `<pad>` and `k` point to `0` - hence the hackiness with the current tokenizer file to make this work, since there isn't a natural way of doing this in `transformers`

Reply:
Again answering with the previus comment from the review I made last week:

> I think @amyeroberts raised very valid questions.
if they pad with k then k is the padding token non?
if <pad> and k both point to 0, we don't really have a way in transformers to make sure this can be done. But I don't see why it should be done since they pad with k! So the padding token is k.
if the problem is that our tokenize methods splits on the special tokens, including padding, but you don't want this to happen, then you can just overwrite tokenize. Given that this model does not seem to use bos nor eos you can probably ignore the lstrip and rstip logics, and just take what you have in _tokenize.

Reply:
Can't we just use the padding token set to `k`?

Reply:
Ah we can't otherwise it will be skipped my bad

Reply:
Gone with this proposed 'hack' - we force the pad token at word boundaries, and then encode any `k`'s to their appropriate id's. Note also that we are unlikely to ever need to decode `input_ids` with a TTS model - we put the `input_ids` in to the model, and get speech out.